File: 611-valid-triangle-number/611-valid-triangle-number.java
class Solution {
    public int triangleNumber(int[] nums) {
        Arrays.sort(nums);
        
        int combs = 0;
        for(int i = 0; i < nums.length - 2; i++){
            int lo = i + 2;
            for(int j = i + 1; j < nums.length - 1; j++){
                int total = nums[i] + nums[j];
                
                int hi = nums.length;
                while(lo < hi){
                    int mid = lo + (hi - lo) / 2;
                    if(nums[mid] >= total)
                        hi = mid;
                    else
                        lo = mid + 1;
                }
                combs += Math.max(0, lo - j - 1);
            }
        }
        return combs;
    }
}

File: 611-valid-triangle-number/611-valid-triangle-number.py
class Solution:
    def triangleNumber(self, nums: List[int]) -> int:
        nums.sort()
        
        n = len(nums)
        
        combs = 0
        for i in range(n - 2):
            lo = i + 2
            for j in range(i + 1, n - 1):
                total = nums[i] + nums[j]
                
                hi = n
                while lo < hi:
                    mid = lo + (hi - lo) // 2
                    if nums[mid] >= total:
                        hi = mid
                    else:
                        lo = lo + 1
                combs += max(0, lo - j - 1)
        return combs